matrix/max_area_of_island.py

```diff
@@ -0,0 +1,126 @@
+# Given an m x n binary matrix grid. An island is a group of 1's
+# (representing land) connected 4-directionally (horizontal or vertical.)
+# You may assume all four edges of the grid are surrounded by water.
+# The area of an island is the number of cells with a value 1 in the island.
+# Return the maximum area of an island in grid.
+# If there is no island, return 0.
+
+
+def is_safe(row: int, col: int, rows: int, cols: int) -> bool:
+    """
+    Checking weather coordinate (row,col) is valid or not.
+
+    >>> is_safe(0, 0, 5, 5)
+    True
+
+    >>> is_safe(-1,-1, 5, 5)
+    False
+    """
+    return not (row < 0 or col < 0 or row >= rows or col >= cols)
+
+
+def dfs(row: int, col: int, ROWS: int, COLS: int, seen: set) -> int:
+    """
+    Returns the current area of the island
+    >>> dfs(0,0, 8,8,set())
+    0
+    """
+    if is_safe(row, col, ROWS, COLS) and (row, col) not in seen and mat[row][col] == 1:
+        seen.add((row, col))
+        return (
+            1
+            + dfs(row + 1, col, ROWS, COLS, seen)
+            + dfs(row - 1, col, ROWS, COLS, seen)
+            + dfs(row, col + 1, ROWS, COLS, seen)
+            + dfs(row, col - 1, ROWS, COLS, seen)
+        )
+    else:
+        return 0
+
+
+def count_max_area(mat: list[list[int]]) -> int:
+    """
+    Finds the area of all islands and returns the maximum area.
+
+    >>> count_max_area([ \
+        [0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0], \
+        [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0], \
+        [0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], \
+        [0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0], \
+        [0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0], \
+        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0], \
+        [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0], \
+        [0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0]])
+    6
+    """
+
+    ROWS = len(mat)
+    COLS = len(mat[0])
+    seen: set = set()
+
+    max_area = 0
+    for row in range(ROWS):
+        for col in range(COLS):
+            if mat[row][col] == 1 and (row, col) not in seen:
+                # Maximizing the area
+                max_area = max(max_area, dfs(row, col, ROWS, COLS, seen))
+    return max_area
+
+
+# Example 1:
+mat = [
+    [0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
+    [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
+    [0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
+    [0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0],
+    [0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0],
+    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
+    [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
+    [0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0],
+]
+
+
+print(count_max_area(mat))  # Output -> 6
```

Comment:
Could you put this (your entire example) inside the `if __name__ == "__main__"` block, or inside a doctest?
This is so if this file is `import`ed, it won't run your example. (Meaning it would only run the example if you directly run the file)

Reply:
yes, it makes sense moving example inside if block